map/src/main/java/openlr/map/utils/BearingDestinationRouting.java

```diff
@@ -0,0 +1,130 @@
+     * @return BearingDestinationRouting
+     */
+    public static BearingDestinationRouting withConfig(Line lrpLine, int bearingDistance, int projectionAlongLine) {
+        return new BearingDestinationRouting(lrpLine, bearingDistance, projectionAlongLine);
+    }
+
+    /**
+     * Verify the relevant end (Start/End node) of the line is not an intersection
+     *
+     * @param connectedLine lines connected to relevant end
+     * @return true: if there is only one line connected to the node
+     * false: if there are more line connected to the node
+     */
+    private boolean isNotIntersection(Iterator<Line> connectedLine) {
+        int counter = 0;
+        while (connectedLine.hasNext()) {
+            connectedLine.next();
+            ++counter;
+        }
+        return (counter == 1);
+    }
+
+    /**
+     * Lines connected to the relevant end of the line
+     *
+     * @param line                parent line
+     * @param inLocationDirection true: if bearing direction is same as the overall location direction
+     * @return if 'inLocationDirection == true' outgoing lines of the end node
+     * else if 'inLocationDirection == false' incoming lines of the start node
+     */
+    private Iterator<Line> calculateNextLineInRoute(Line line, boolean inLocationDirection) {
+        if (inLocationDirection) {
+            return line.getNextLines();
+        } else {
+            return line.getPrevLines();
+        }
+    }
+
+    /**
+     * @param line                line on which bearing destination exist
+     * @param offset              distance from the relevant end, based on the location direction, of the line to the bearing destination
+     * @param inLocationDirection true: if bearing direction is same as the overall location direction
+     * @return bearing destination coordinates
+     */
+    private GeoCoordinates calculateBearingDestinationOnLine(Line line, int offset, boolean inLocationDirection) {
+        if (inLocationDirection) {
+            return line.getGeoCoordinateAlongLine(offset);
+        } else {
+            return line.getGeoCoordinateAlongLine(line.getLineLength() - offset);
+        }
+
+    }
+
+    /**
+     * @param lrpLine                line on which lrp is placed
+     * @param lengthCoveredOnLrpLine distance along the line from the point on the line to which the LRP is projected to the relevant end based on bearing direction
+     * @param inLocationDirection    true: if bearing direction is same as the overall location direction
+     * @return Coordinates of the bearing destination
+     */
+    private GeoCoordinates getBearingDestination(Line lrpLine, int lengthCoveredOnLrpLine, boolean inLocationDirection) {
+        Line currentLine = lrpLine;
+        int lengthCovered = lengthCoveredOnLrpLine;
+        while (bearingDistance >= lengthCovered && isNotIntersection(calculateNextLineInRoute(currentLine, inLocationDirection))) {
```

Comment:
In line 95 and 96 the next lines of the current line are being fetched. Depending on the map implementation this might be an expensive operation.  I think it's better to just fetch them once.  For example, if you had a method

```
private List<Line> getNextLines(Line current, boolean inLocationDirection) {
    Iterator<Line> iterator = inLocationDirection ? current.getNextLines() : current.getPrevLines();
    return StreamSupport
            .stream(Spliterators.spliteratorUnknownSize(iterator, 0), false)
            .collect(Collectors.toList());
}
``` 

you could then change your loop to this

```
Line currentLine = lrpLine;
int lengthCovered = lengthCoveredOnLrpLine;
List<Line> nextLines = getNextLines(currentLine, inLocationDirection);
while (bearingDistance >= lengthCovered && nextLines.size() == 1) {
    currentLine = nextLines.get(0);
    lengthCovered += currentLine.getLineLength();
    nextLines = getNextLines(currentLine, inLocationDirection);
}
```

and then you wouldn't need the 2 utility methods.  I think it would be more efficient and more readable.

Reply:
we need to explore the network only if the bearing distance cannot be covered on the lrp line.

Reply:
482ec5f58091993383b0ec513ceafb5223fa03d7